pkg/token/serviceprincipaltoken_test.go

```diff
@@ -0,0 +1,232 @@
+		}
+
+		if strings.Contains(i.Request.Body, "client_assertion") {
+			i.Request.Body = `client_assertion=[REDACTED]&client_assertion_type=urn%3Aietf%3Aparams%3Aoauth%3Aclient-assertion-type%3Ajwt-bearer&client_id=[REDACTED]&client_info=1&grant_type=client_credentials&scope=[REDACTED]%2F.default+openid+offline_access+profile`
+		}
+
+		if strings.Contains(i.Response.Body, "access_token") {
+			i.Response.Body = `{"token_type":"Bearer","expires_in":86399,"ext_expires_in":86399,"access_token":"[REDACTED]"}`
+		}
+		return nil
+	}
+	rec.AddHook(hook, recorder.BeforeSaveHook)
+
+	rec.SetMatcher(customMatcher)
+	rec.SetReplayableInteractions(true)
+
+	return rec, rec.GetDefaultClient()
+}
+
+func customMatcher(r *http.Request, i cassette.Request) bool {
+	id := os.Getenv(tenantID)
+	if id == "" {
+		id = "00000000-0000-0000-0000-000000000000"
+	}
+	switch os.Getenv(vcrMode) {
+	case vcrModeRecordOnly:
+	default:
+		r.URL.Path = strings.ReplaceAll(r.URL.Path, id, tenantID)
+	}
+	return cassette.DefaultMatcher(r, i)
+}
+
+// Get go-vcr record mode from enviroment variable
+func getVCRMode() recorder.Mode {
+	switch os.Getenv(vcrMode) {
+	case vcrModeRecordOnly:
+		return recorder.ModeRecordOnly
+	default:
+		return recorder.ModeReplayOnly
+	}
+}
+
+func TestServicePrincipalLoginVCR(t *testing.T) {
+	testCase := []struct {
+		cassetteName  string
+		description   string
+		p             *servicePrincipalToken
+		expectedError error
+	}{
+		{
+			cassetteName: "BadSecretVCR",
+			description:  "Test using incorrect secret value",
+			p: &servicePrincipalToken{
+				clientID:     os.Getenv(clientID),
+				clientSecret: badSecret,
+				resourceID:   os.Getenv(resourceID),
+				tenantID:     os.Getenv(tenantID),
+			},
+			expectedError: fmt.Errorf("ClientSecretCredential authentication failed"),
+		},
+		{
+			cassetteName: "SecretTokenVCR",
+			description:  "Test using service principal secret value to get token",
+			p: &servicePrincipalToken{
+				clientID:     os.Getenv(clientID),
+				clientSecret: os.Getenv(clientSecret),
+				resourceID:   os.Getenv(resourceID),
+				tenantID:     os.Getenv(tenantID),
+			},
+			expectedError: nil,
+		},
+		{
+			cassetteName: "CertTokenVCR",
+			description:  "Test ",
+			p: &servicePrincipalToken{
+				clientID:           os.Getenv(clientID),
+				clientCert:         os.Getenv(clientCert),
+				clientCertPassword: os.Getenv(clientCertPass),
+				resourceID:         os.Getenv(resourceID),
+				tenantID:           os.Getenv(tenantID),
+			},
+			expectedError: nil,
+		},
+	}
+
+	for _, tc := range testCase {
+		t.Run(tc.description, func(t *testing.T) {
+			vcrRecorder, httpClient := getVCRHttpClient(fmt.Sprintf("testdata/%s", tc.cassetteName))
+
+			clientOpts := azcore.ClientOptions{
+				Cloud:     cloud.AzurePublic,
+				Transport: httpClient,
+			}
+			if tc.p.clientID == "" {
+				tc.p.clientID = clientID
+			}
+			if tc.p.clientSecret == "" && strings.Contains(tc.cassetteName, "Secret") {
```

Comment:
providing a flag in the test case struct would be more explicit and easier to read. using cassette name as convention is too vague.

Reply:
Fixed